core/src/main/scala/dagr/core/execsystem/TaskManager.scala

```diff
@@ -327,20 +328,21 @@ class TaskManager(taskManagerResources: SystemResources = TaskManagerDefaults.de
     * @return for each completed task, a map from a task identifier to a tuple of the exit code and the status of the `onComplete` method
     */
   private def updateCompletedTasks(): Map[TaskId, (Int, Boolean)] = {
-    val completedTasks: Map[TaskId, (Int, Boolean)] = taskExecutionRunner.completedTasks()
-    val emptyTasks = graphNodesInStateFor(GraphNodeState.NO_PREDECESSORS).filter(_.task.isInstanceOf[Task.EmptyTask])
-
-    emptyTasks.foreach { node =>
+    // Get the tasks that are not eligible to be scheduled/executed, and have no predecessors (can be completed).
+    val toCompleteTasks = graphNodesInStateFor(GraphNodeState.NO_PREDECESSORS).filterNot(_.task.isInstanceOf[UnitTask])
+    toCompleteTasks.foreach { node =>
       node.taskInfo.status = TaskStatus.SUCCEEDED
       processCompletedTask(node.taskId)
-      logger.debug("updateCompletedTasks: empty task [" + node.task.name + "] completed")
+      logger.debug("updateCompletedTasks: to-complete task [" + node.task.name + "] completed")
     }
 
+    // Get the tasks that have been executed to completion.
+    val completedTasks: Map[TaskId, (Int, Boolean)] = taskExecutionRunner.completedTasks()
```

Comment:
Why not just have a single loop with:
 `(toCompleteTasks.map(_.task_id) ++ completedTasks.keys).foreach { taskId => ... }`
I get that the log messages are different, but you could just make them the same and I think it would be clearer that you're not actually doing different handling.